### PR TITLE
feat(filetype.lua): add support for patch files

### PR DESCRIPTION
**Pull request by SanchayanMaity**

### CODE REVIEW

File: runtime/lua/vim/filetype.lua [hunk 1]
```diff
@@ -409,6 +409,7 @@ local extension = {
   opam = "opam",
   ["or"] = "openroad",
   ora = "ora",
+  patch = "gitsendemail",
```
Comment (gpanders) [inline runtime/lua/vim/filetype.lua:412]: This needs to be a function that checks the first line:

```vim
au BufNewFile,BufRead *.patch
	\ if getline(1) =~ '^From [0-9a-f]\{40\} Mon Sep 17 00:00:00 2001$' |
	\   setf gitsendemail |
	\ else |
	\   setf diff |
	\ endif
``

**Comment by clason** [comment]
Can you add the similar pattern for `git` files from https://github.com/neovim/neovim/pull/16953#discussion_r779564301?

We also need to port the tests for these.

**Comment by SanchayanMaity** [comment]
@clason Will add the git pattern. Can you point me to where the tests go/existing test?

**Comment by clason** [comment]
Sure, look at the linked PR (`test/functional/lua/filetype_spec.lua`)!

**Comment by SanchayanMaity** [comment]
Not sure how to write test for `patch`. It relies on reading the first line from the file. In `test_filetype.vim` they do `writefile` calls which is understandable.

File: runtime/lua/vim/filetype.lua [hunk 2]
```diff
@@ -770,6 +770,14 @@ local extension = {
   mm = function() vim.fn["dist#ft#FTmm"]() end,
   mms = function() vim.fn["dist#ft#FTmms"]() end,
   p = function() vim.fn["dist#ft#FTprogress_pascal"]() end,
+  patch = function(path, bufnr)
+    local firstline = vim.fn.getline(1)
```
Comment (gpanders) [inline runtime/lua/vim/filetype.lua:774]: Use the `getline` helper function from this file:

```suggestion
    local firstline = getline(bufnr, 1)
```

File: runtime/lua/vim/filetype.lua [hunk 3]
```diff
@@ -770,6 +770,14 @@ local extension = {
...
+    if string.find(firstline, "From ") and string.find(firstline, "Mon Sep 17 00:00:00 2001") then
```
Comment (gpanders) [inline runtime/lua/vim/filetype.lua:775]: I don't understand why this pattern is used, but it's what filetype.vim does so ¯\\\_(ツ)\_/¯ 

```suggestion
    if string.find(firstline, "^From " .. string.rep("%x", 40) .. "+ Mon Sep 17 00:00:00 2001$") then
```

File: runtime/lua/vim/filetype.lua [hunk 4]
```diff
@@ -1334,6 +1342,12 @@ local pattern = {
   ["zlog.*"] = starsetf('zsh'),
   ["zsh.*"] = starsetf('zsh'),
   ["ae%d+%.txt"] = 'mail',
+  [".*%.git/.*"] = function(path, bufnr)
+    local firstline = vim.fn.getline(1)
```
Comment (gpanders) [inline runtime/lua/vim/filetype.lua:1346]:
```suggestion
    local firstline = getline(bufnr, 1)
```

**Comment by SanchayanMaity** [comment]
Currently, this is what I came up with for the tests but tests which depend on the written file content are all failing. Seems I am missing something in my understanding.

**Comment by gpanders** [comment]
@SanchayanMaity Thanks for doing that work, but I think we're going to go in a different direction with the filetype.lua tests (see https://github.com/neovim/neovim/pull/17003) which means porting the tests to `filetype_spec.lua` won't be necessary. I'm sorry I asked you to do that for nothing, hopefully you didn't spend too much time on it!

Your changes look good to me upon inspection, let me test them against the existing `test_filetype.vim` and if they pass then we can merge this 👍

**Comment by gpanders** [comment]
Added to #16978. Thanks!

**Comment by SanchayanMaity** [comment]
@gpanders Cool. No worries.

**Pull request closed without merging**